### ISSUE TITLE
Cannot connect to an exposed nodeport from host system. 

### ISSUE THREAD

**Issue by NickPaul41** [crc-org/crc#3393]
### General information

  * OS:  Windows
  * Hypervisor: Hyper-V 
  * Did you run `crc setup` before starting it (Yes/No)? yes
  * Running CRC on: Laptop 

## CRC version

CRC version: 2.10.1+7e7f6b2d
OpenShift version: 4.11.7
Podman version: 4.2.0

## CRC status

DEBU CRC version: 2.10.1+7e7f6b2d
DEBU OpenShift version: 4.11.7
DEBU Podman version: 4.2.0
DEBU Running 'crc status'
DEBU Checking file: C:\Users\nick\.crc\machines\crc\.crc-exist
DEBU Checking file: C:\Users\nick\.crc\machines\crc\.crc-exist
DEBU Running 'Hyper-V\Get-VM crc | Select-Object -ExpandProperty State'
DEBU Running SSH command: df -B1 --output=size,used,target /sysroot | tail -1
DEBU Using ssh private keys: [C:\Users\nick\.crc\machines\crc\id_ecdsa C:\Users\nick\.crc\cache\crc_hyperv_4.11.7_amd64\id_ecdsa_crc]
DEBU SSH command results: err: <nil>, output: 32737570816 23658688512 /sysroot
CRC VM:          Running
OpenShift:       Running (v4.11.7)
Podman:
Disk Usage:      23.66GB of 32.74GB (Inside the CRC VM)
Cache Usage:     22.02GB
Cache Directory: C:\Users\nick\.crc\cache

## CRC config

- consent-telemetry                     : no
- cpus                                  : 10
- enable-cluster-monitoring             : true
- host-network-access                   : true
- memory                                : 22950
- 
## Host Operating System
Host Name:                 VOID
OS Name:                   Microsoft Windows 11 Pro
OS Version:                10.0.22000 N/A Build 22000
OS Manufacturer:           Microsoft Corporation
OS Configuration:          Standalone Workstation
OS Build Type:             Multiprocessor Free
Registered Owner:          nwpaul
Registered Organization:
Product ID:                00330-80000-00000-AA651
Original Install Date:     5/4/2022, 3:41:01 PM
System Boot Time:          10/19/2022, 6:56:38 AM
System Manufacturer:       Gigabyte Technology Co., Ltd.
System Model:              X399 DESIGNARE EX
System Type:               x64-based PC
Processor(s):              1 Processor(s) Installed.
                           [01]: AMD64 Family 23 Model 1 Stepping 1 AuthenticAMD ~3400 Mhz
BIOS Version:              American Megatrends Inc. F13a, 11/30/2021
Windows Directory:         C:\Windows
System Directory:          C:\Windows\system32
Boot Device:               \Device\HarddiskVolume2
System Locale:             en-us;English (United States)
Input Locale:              en-us;English (United States)
Time Zone:                 (UTC-08:00) Pacific Time (US & Canada)
Total Physical Memory:     65,415 MB
Available Physical Memory: 21,641 MB
Virtual Memory: Max Size:  75,143 MB
Virtual Memory: Available: 13,792 MB
Virtual Memory: In Use:    61,351 MB
Page File Location(s):     C:\pagefile.sys
Domain:                    WORKGROUP
Logon Server:              \\VOID
Hotfix(s):                 5 Hotfix(s) Installed.
                           [01]: KB5017024
                           [02]: KB5012170
                           [03]: KB5018418
                           [04]: KB5017850
                           [05]: KB5018291
Network Card(s):           5 NIC(s) Installed.
                           [01]: Intel(R) Dual Band Wireless-AC 8265
                                 Connection Name: Wi-Fi
                                 Status:          Media disconnected
                           [02]: Intel(R) I211 Gigabit Network Connection
                                 Connection Name: Ethernet
                                 Status:          Media disconnected
                           [03]: Intel(R) I211 Gigabit Network Connection
                                 Connection Name: Ethernet 2
                                 DHCP Enabled:    Yes
                                 DHCP Server:     192.168.0.1
                                 IP address(es)
                                 [01]: 192.168.0.41
                                 [02]: fe80::bcc5:b6c2:cf71:8418
                           [04]: Bluetooth Device (Personal Area Network)
                                 Connection Name: Bluetooth Network Connection
                                 Status:          Media disconnected
                           [05]: PANGP Virtual Ethernet Adapter
                                 Connection Name: Ethernet 3
                                 DHCP Enabled:    No
                                 IP address(es)
                                 [01]: 10.250.249.103
                                 [02]: fc00::1:3e1c
Hyper-V Requirements:      A hypervisor has been detected. Features required for Hyper-V will not be displayed.

### Steps to reproduce

  1.  Have a pod expose a port 8081 (non-http)
  2.  Expose it as a nodeport 
  3. oc expose deployment myDeployment --name=myDeployment  --type=NodePort --port=8081
  4. From the hosting system connect to port 8081

### Expected
Connection successful and the ability to communicate with the nodeport.

### Actual
Connection Timeout

**Comment by praveenkumar** [comment]
@NickPaul41 for windows are using the usermode network stack and also a container which update the route url to host's`/etc/hosts` if you want to access that service on host better to expose this service to route instead of nodeport.
```
$ oc expose deployment myDeployment --name=myDeployment --port=8081
$ oc expose svc myDeployment
$ oc get routes
```

routes command provide you `HOST` field which you can use to access it on the host with specified port.

**Comment by NickPaul41** [comment]
I'm not using HTTP, but a tcp socket connection. I've tried using routes before but they are setup for HTTP and I get a 400 bad request when making the tcp connection. In a OpenShift cluster I expose a node port and then connect to it.  I'm assuming node ports are not available though crc. Is there another way to open a socket connection (non-http) between crc and host?

**Comment by mltucker** [comment]
When using TCP, use `oc port-forward <pod> 8888:5000`

**Comment by NickPaul41** [comment]
Thanks @mltucker that does work. Is there a more permanent way to set this up? 

I have a few services to interact with and the pod names change after recreation or rebooting my PC.

**Comment by anjannath** [comment]
Maybe the routes controller can be extended to also watch for nodePorts alongside routes and the daemon can then automatically expose the port on the host using gvproxy?

**Comment by raynay-r** [comment]
I created a controller analogous to the routes controller which handles this. You can take a look here: https://github.com/raynay-r/crc-node-port-controller

Would you be interested in adding this controller to the CRC distribution?

**Comment by cfergeau** [comment]
Can you provide this as an addition to the routes-controller (even though naming may need to evolve :) It's probably better to have both routes and node ports in the same codebase instead of keeping both separate?

**Comment by raynay-r** [comment]
Yes, i can provide a PR for the routes-controller, but will probably take few weeks until I get to it.

**Comment by praveenkumar** [comment]
@NickPaul41 We have already added this as part of route controller https://github.com/crc-org/routes-controller/pull/12 which should fix it. Please try latest version of OpenShift Local and create a new issue if this is still not fixed.

**Comment by cfergeau** [comment]
We haven't made a routes-controller release since that was merged, so it's likely this isn't fully fixed yet :(

**Comment by praveenkumar** [comment]
> We haven't made a routes-controller release since that was merged, so it's likely this isn't fully fixed yet :(

As per yesterday discussion we did just not bump up the release so it should be fixed right?

**Comment by cfergeau** [comment]
Yes
